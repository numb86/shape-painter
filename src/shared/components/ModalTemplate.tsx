import React, {useEffect, useCallback} from 'react';
import styled from 'styled-components';
import dialogPolyfill from 'dialog-polyfill';

type Props = {
  modalRef: React.MutableRefObject<HTMLDialogElement | null>;
  closeModal: (
    event:
      | React.MouseEvent<HTMLDialogElement | HTMLButtonElement, MouseEvent>
      | React.FormEvent<HTMLFormElement>
  ) => void;
  headerTitle: string;
  layoutSetting: {
    modalWidth: number;
    mainOneSidePadding: number;
  };
};

export const ModalTemplate: React.FC<Props> = ({
  modalRef,
  closeModal,
  headerTitle,
  layoutSetting,
  children,
}) => {
  useEffect(() => {
    if (process.env.NODE_ENV !== 'test') {
      if (modalRef.current && !modalRef.current.showModal) {
        dialogPolyfill.registerDialog(modalRef.current);
      }
    }
  }, [modalRef]);

  const stopPropagation = useCallback(
    (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
      e.stopPropagation();
    },
    []
  );

  return (
    <Modal
      onClick={closeModal}
      ref={modalRef}
      width={layoutSetting.modalWidth}
      data-testid="modal"
    >
      <ModalBody onClick={stopPropagation} data-testid="modal-body">
        <Header data-testid="header">{headerTitle}</Header>
        <Main
          oneSidePadding={layoutSetting.mainOneSidePadding}
          data-testid="main"
        >
          {children}
        </Main>
      </ModalBody>
    </Modal>
  );
};

const BORDER_RADIUS = 30;

const Modal = styled.dialog<{width: number}>`
  /* For Native */
  &::backdrop {
    background-color: #000;
    opacity: 0.3;
  }

  /* For Polyfill */
  + .backdrop {
    background-color: #000;
    opacity: 0.3;
    position: absolute;
    top: 0;
    width: 100%;
    height: 100%;
  }
  &:not([open]) {
    display: none;
  }
  height: fit-content;
  position: absolute;
  left: 0;
  right: 0;
  margin: auto;

  width: ${(props) => props.width}px;
  top: 70px;
  border: none;
  padding: 0;
  background: transparent;
  color: #000;
`;

const ModalBody = styled.div``;

const Header = styled.header`
  background-color: #fff;
  border-bottom: 2px solid rgb(58, 171, 210);
  height: 45px;
  padding-top: 20px;
  border-top-left-radius: ${BORDER_RADIUS}px;
  border-top-right-radius: ${BORDER_RADIUS}px;
  font-size: 23px;
  text-align: center;
`;

const Main = styled.main<{oneSidePadding: number}>`
  background-color: #fff;
  padding: 30px ${(props) => props.oneSidePadding}px;
  border-bottom-left-radius: ${BORDER_RADIUS}px;
  border-bottom-right-radius: ${BORDER_RADIUS}px;
`;
