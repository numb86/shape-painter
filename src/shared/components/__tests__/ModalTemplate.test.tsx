import React from 'react';
import {render, fireEvent, screen} from '@testing-library/react';
import 'jest-styled-components';
import assert from 'power-assert';

import {ModalTemplate} from '../ModalTemplate';

describe('ModalTemplate', () => {
  let closeModal: jest.Mock<any, any>;
  const HEADER_TITLE = 'headerText';
  const MAIN_CONTENT = 'abc';
  let firstChild: ChildNode;

  beforeEach(() => {
    closeModal = jest.fn();

    const {container} = render(
      <ModalTemplate
        modalRef={React.createRef()}
        closeModal={closeModal}
        headerTitle={HEADER_TITLE}
        layoutSetting={{
          modalWidth: 200,
          mainOneSidePadding: 55,
        }}
      >
        {MAIN_CONTENT}
      </ModalTemplate>
    );

    firstChild = container.firstChild!;
  });

  it('Clicking Modal calls props.closeModal', () => {
    expect(closeModal).toHaveBeenCalledTimes(0);
    fireEvent.click(screen.getByTestId('modal'));
    expect(closeModal).toHaveBeenCalledTimes(1);
  });

  it('Clicking Modal content does not call props.closeModal', () => {
    expect(closeModal).toHaveBeenCalledTimes(0);
    fireEvent.click(screen.getByTestId('modal-body'));
    expect(closeModal).toHaveBeenCalledTimes(0);
  });

  it('props.headerTitle becomes content of header', () => {
    assert.strictEqual(screen.getByTestId('header').textContent, HEADER_TITLE);
  });

  it('props.children appears in main', () => {
    assert.strictEqual(screen.getByTestId('main').textContent, MAIN_CONTENT);
  });

  it('Contents of props.layoutSetting are reflected in the style', () => {
    expect(firstChild).toMatchSnapshot();
  });
});
