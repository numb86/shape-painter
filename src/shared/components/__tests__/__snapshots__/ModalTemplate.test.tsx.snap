// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalTemplate Contents of props.layoutSetting are reflected in the style 1`] = `
.c0 {
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  position: absolute;
  left: 0;
  right: 0;
  margin: auto;
  width: 200px;
  top: 70px;
  border: none;
  padding: 0;
  background: transparent;
  color: #000;
}

.c0::backdrop {
  background-color: #000;
  opacity: 0.3;
}

.c0 + .backdrop {
  background-color: #000;
  opacity: 0.3;
  position: absolute;
  top: 0;
  width: 100%;
  height: 100%;
}

.c0:not([open]) {
  display: none;
}

.c1 {
  background-color: #fff;
  border-bottom: 2px solid rgb(58,171,210);
  height: 45px;
  padding-top: 20px;
  border-top-left-radius: 30px;
  border-top-right-radius: 30px;
  font-size: 23px;
  text-align: center;
}

.c2 {
  background-color: #fff;
  padding: 30px 55px;
  border-bottom-left-radius: 30px;
  border-bottom-right-radius: 30px;
}

<dialog
  class="c0"
  data-testid="modal"
  width="200"
>
  <div
    class=""
    data-testid="modal-body"
  >
    <header
      class="c1"
      data-testid="header"
    >
      headerText
    </header>
    <main
      class="c2"
      data-testid="main"
    >
      abc
    </main>
  </div>
</dialog>
`;
